### PR TITLE
upgrade: Enable node upgrade status api

### PR DESCRIPTION
**Pull request by MaximilianMeister**
this lists all upgraded and all non upgraded nodes

## NOTE
~~`crowbarctl` PR will follow~~

here it is https://github.com/crowbar/crowbar-client/pull/147

## example response
```bash
{
  "upgraded": [
  ],
  "not_upgraded": [
    "d52-54-77-77-77-03.emil.suse.de",
    "d52-54-77-77-77-01.emil.suse.de",
    "d52-54-77-77-77-02.emil.suse.de",
    "d52-54-77-77-77-04.emil.suse.de"
  ]
}
```

### CODE REVIEW

**Comment by jsuchome** [comment]
Do we really need a backport here?

**Comment by MaximilianMeister** [comment]
> Do we really need a backport here?

yes, as we will have the same `crowbarctl` on cloud6

**Comment by MaximilianMeister** [comment]
[(5) Allow user to choose which compute nodes to upgrade](https://trello.com/c/IZ8W5tQz/224-5-allow-user-to-choose-which-compute-nodes-to-upgrade)

File: crowbar_framework/app/models/api/upgrade.rb [hunk 1]
```diff
@@ -23,6 +23,13 @@ def status
         ::Crowbar::UpgradeStatus.new.progress
       end
 
+      def node_status
+        {
+          upgraded: ::Node.all.reject(&:admin?).select(&:ready_after_upgrade?).map(&:name),
+          not_upgraded: ::Node.find("state:crowbar_upgrade").map(&:name)
```
Comment (skazi0) [inline crowbar_framework/app/models/api/upgrade.rb]: are we sure that this will give us all nodes which are not upgraded yet?

Comment (MaximilianMeister) [inline crowbar_framework/app/models/api/upgrade.rb]: i guess only when it's executed after the `prepare` step. @jsuchome is this an issue?

Comment (jsuchome) [inline crowbar_framework/app/models/api/upgrade.rb]: Hm, you are right that before `prepare` (heck, even before `nodes`) it does not make sense.
Maybe we could just check for the current step here?

Comment (MaximilianMeister) [inline crowbar_framework/app/models/api/upgrade.rb]: should be fixed now

Comment (skazi0) [inline crowbar_framework/app/models/api/upgrade.rb]: what if something gets out-of-sync somehow? iirc I had some cases where node switched to 'no-data' and then when it returned it was 'ready' not 'upgrade'... but I might be wrong.

Comment (jsuchome) [inline crowbar_framework/app/models/api/upgrade.rb]: That "no data" state should be covered by `:ready_after_upgrade?` I think

File: crowbar_framework/app/models/api/upgrade.rb [hunk 2]
```diff
@@ -23,6 +23,19 @@ def status
         ::Crowbar::UpgradeStatus.new.progress
       end
 
+      def node_status
+        not_upgraded = if ::Crowbar::UpgradeStatus.new.pending?(:prepare) || \
```
Comment (jsuchome) [inline crowbar_framework/app/models/api/upgrade.rb]: Seems like `passed(step)?` method in the library would be handy...

Comment (MaximilianMeister) [inline crowbar_framework/app/models/api/upgrade.rb]: done

**Pull request merged**